pandas/core/reshape.py

```diff
@@ -232,19 +226,31 @@ def get_new_index(self):
         # construct the new index
         if len(self.new_index_levels) == 1:
             lev, lab = self.new_index_levels[0], result_labels[0]
-            if not (lab == -1).any():
-                return lev.take(lab)
-
-            vals = np.insert(lev.astype('object'), len(lev),
-                             _get_na_value(lev.dtype.type)).take(lab)
-
-            return lev._shallow_copy(vals)
+            return _make_new_index(lev, lab) \
+                    if (lab == -1).any() else lev.take(lab)
 
         return MultiIndex(levels=self.new_index_levels,
                           labels=result_labels,
                           names=self.new_index_names,
                           verify_integrity=False)
 
+
+def _make_new_index(lev, lab):
+    from pandas.core.index import Index, _get_na_value
+
+    nan = _get_na_value(lev.dtype.type)
+    vals = lev.values.astype('object')
+    vals = np.insert(vals, 0, nan) if lab is None else \
+           np.insert(vals, len(vals), nan).take(lab)
+
+    try:
+        vals = vals.astype(lev.dtype, subok=False, copy=False)
+    except ValueError:
+        return Index(vals, **lev._get_attributes_dict())
+
+    return lev._shallow_copy(vals)
+
```

Comment:
OK, I think adding the following  code just before the ``try`` fixes ``_make_new_index(lev, lab)``. I will include it in #9023.
```
    if lev.dtype.type == np.datetime64:
        nan_indices = [0] if lab is None else (np.array(lab) == -1)
        vals[nan_indices] = None
```